sdks/go/pkg/beam/core/runtime/exec/callers.tmpl

```diff
@@ -0,0 +1,64 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+    "reflect"
+
+  	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+  	"github.com/apache/beam/sdks/go/pkg/beam/core/util/reflectx"
+)
+
+// TODO(herohde) 1/4/2018: Potential targets for type-specialization include simple predicate,
+// dofn and combiner forms.
+//
+//   (1) func(X) bool
+//   (2) func(X) error
+//   (3) func(X, X) X
+//   (4) func(ctx?, X) (X, error?)
+//   (5) func(ctx?, X, func(X)) (error?)
+//
+// For now, we just do #3.
+
+func init() {
+{{- range $x := .X}}
+    reflectx.RegisterCaller(reflect.TypeOf((*func ({{$x.Type}}, {{$x.Type}}){{$x.Type}})(nil)).Elem(), callMaker{{$x.Name}}M)
+{{- end}}
+}
+
+{{range $x := .X}}
+type native{{$x.Name}}MCaller struct {
+	fn    func ({{$x.Type}}, {{$x.Type}}){{$x.Type}}
+}
+
+func callMaker{{$x.Name}}M(fn interface{}) reflectx.Caller {
```

Comment:
<!--new_thread; commit:5ff88b0ad86fa84eb8323c61651be06916571224; resolved:0-->
Why M? Merge?

If so, consider updating the TODO header you've nicely got to indicate what these are for specifically.

Consider adding a comment above this struct (the loop I mean, not per struct) and it's methods saying it's to handle Binary Merge combines. Templates should be readable and documented where they won't balloon the size.

(1) Predicate
(2) ??? predicate that returns error?
(3) Binary Merge
(4) & (5) ???? Do Fns?

Reply:
Yes. M for Merge. Clarified the comment.